src/main/java/io/numaproj/pulsar/consumer/PulsarSource.java

```
@@ -7,139 +7,119 @@
+                pulsarConsumerManager.removeConsumer();
+                consumer = pulsarConsumerManager.getOrCreateConsumer(request.getCount(),
+                        request.getTimeout().toMillis());
+                batchMessages = consumer.batchReceive();
+            }
+
+            if (batchMessages == null || batchMessages.size() == 0) {
+                return;
+            }
 
-                log.info("Consumed Pulsar message: {}", new String(pMsg.getValue()));
+            // Process each message in the batch.
+            for (org.apache.pulsar.client.api.Message<byte[]> pMsg : batchMessages) {
+                String msgId = pMsg.getMessageId().toString();
+                log.info("Consumed Pulsar message [id: {}]: {}", pMsg.getMessageId(),
+                        new String(pMsg.getValue(), StandardCharsets.UTF_8));
 
-                byte[] offsetBytes = pMsg.getMessageId().toString().getBytes(StandardCharsets.UTF_8);
+                byte[] offsetBytes = msgId.getBytes(StandardCharsets.UTF_8);
                 Offset offset = new Offset(offsetBytes);
 
-                Map<String, String> headers = new HashMap<>();
-                headers.put("pulsarMessageId", pMsg.getMessageId().toString());
+                HashMap<String, String> headers = new HashMap<>();
+                headers.put("pulsarMessageId", msgId);
 
                 Message message = new Message(pMsg.getValue(), offset, Instant.now(), headers);
-
                 observer.send(message);
-                messages.put(pMsg.getMessageId().toString(), pMsg);
 
-            } catch (PulsarClientException e) {
-                log.error("Failed to receive message from Pulsar", e);
-                return;
+                messagesToAck.put(msgId, pMsg);
             }
+        } catch (PulsarClientException e) {
+            log.error("Failed to get consumer or receive messages from Pulsar", e);
```

Please re-throw the exception after logging.